streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -770,7 +824,7 @@ public void onAssignment(final Assignment assignment) {
                 usedSubscriptionMetadataVersion = leaderSupportedVersion;
             }
 
-            versionProbingFlag.set(true);
+            assignmentErrorCode.set(Error.VERSION_PROBING.code);
```

Comment:
As we are merging the two scenarios to use the error code, we should let the leader to set the error code in the version probing case as well as setting the `receivedAssignmentMetadataVersion` in the assignment. And then we only need to do the logic in line 796 above, and do not need to set it in line 827 any more.

Reply:
Currently the version probing works as the following:

1. when leader receives the subscription info encoded with a higher version that it can understand (e.g. the leader is on version 3, while one of the subscription received is encode with version 4), it will send back an empty assignment with the assignment encoded with version 3, and also `latestSupportedVersion` set to 3.

2. when the member receives the assignment, it checks if `latestSupportedVersion` is smaller than the version it used for encoding the sent subscription (i.e. the above logic). If it is smaller, then it means that leader cannot understand, in this case, version 4. It will then set the flag and then re-subscribe but with a down-graded encoding format of version 3.

NOW with your PR, we can let leader to clearly communicate this error via the error code, and upon receiving the assignment, if the error code is `VERSION_PROBING`, then the member can immediately know what happens, and hence can simplify the above logic. Does that make sense? also cc @mjsax

Reply:
In the example given above, the gap in subscription info versions between leader and the member is 1.
Is the expectation that when the gap is > 1, one round trip is reduced for version probing ?

The version probing error code currently is hard coded and not correlated with the actual gap.

I wonder if the optimization can be done in another JIRA.

Reply:
@guozhangwang For future version this might work. However, if we upgrade from 2.0 to 2.1 with version bump from 3 -> 4, the old leader is on version 3 and cannot encode the version probing via the error flag.

As we are stuck with older version 3 metadata, I am not sure if we gain a lot if we change the logic, as we still need the current code anyway.